pip/req/req_file.py

```diff
@@ -1,137 +1,284 @@
-            index_url = _remove_prefixes(line, '-i', '--index-url')
-            if finder:
-                finder.index_urls = [index_url]
-        elif line.startswith('--extra-index-url'):
-            line = _remove_prefix(line, '--extra-index-url')
-            if finder:
-                finder.index_urls.append(line)
-        elif line.startswith('--use-wheel'):
-            # Default in 1.5
-            pass
-        elif line.startswith('--no-use-wheel'):
-            if finder:
+                # Relative to an URL.
+                req_url = urllib_parse.urljoin(filename, value)
+            elif not _scheme_re.search(value):
+                req_dir = os.path.dirname(filename)
+                req_url = os.path.join(os.path.dirname(filename), value)
+            # TODO: Why not use `comes_from='-r {} (line {})'` here as well?
+            parser = parse_requirements(req_url, finder, comes_from, options, session)
+            for req in parser:
+                yield req
+
+        # ---------------------------------------------------------------------
+        elif linetype == FLAG:
+            if not finder:
+                continue
+
+            if finder and value == '--no-use-wheel':
                 finder.use_wheel = False
-        elif line.startswith('--no-index'):
-            if finder:
+            elif value == '--no-index':
                 finder.index_urls = []
-        elif line.startswith("--allow-external"):
-            line = _remove_prefix(line, '--allow-external')
-            if finder:
-                finder.allow_external |= set([normalize_name(line).lower()])
-        elif line.startswith("--allow-all-external"):
-            if finder:
+            elif value == '--allow-all-external':
                 finder.allow_all_external = True
-        # Remove in 7.0
-        elif line.startswith("--no-allow-external"):
-            pass
-        # Remove in 7.0
-        elif line.startswith("--no-allow-insecure"):
-            pass
-        # Remove after 7.0
-        elif line.startswith("--allow-insecure"):
-            line = _remove_prefix(line, '--allow-insecure')
-            if finder:
-                finder.allow_unverified |= set([normalize_name(line).lower()])
-        elif line.startswith("--allow-unverified"):
-            line = _remove_prefix(line, '--allow-unverified')
-            if finder:
+
+        # ---------------------------------------------------------------------
+        elif linetype == OPTION:
+            if not finder:
+                continue
+
+            opt, value = value
+            if opt == '-i' or opt == '--index-url':
+                finder.index_urls = [value]
+            elif opt == '--extra-index-url':
+                finder.index_urls.append(value)
+            elif opt == '--allow-external':
+                finder.allow_external |= set([normalize_name(value).lower()])
+            elif opt == '--allow-insecure':
+                # Remove after 7.0
```

Comment:
The options that were there for compatibility reasons were moved to the top of `req_file.py`:

```python
# Encountering any of these is a no-op.
parser_compat = set([
    '-Z', '--always-unzip',
    '--use-wheel',          # Default in 1.5
    '--no-allow-external',  # Remove in 7.0
    '--no-allow-insecure',  # Remove in 7.0
])
```

There don't seem to be any tests that check if obsolete options are ignored at the moment.